Update PinJoint2D API with angle limits and motor speed

Ughuuu opened the pull request:
Closes: https://github.com/godotengine/godot-proposals/issues/7646

Ughuuu committed:
Update PinJoint2D API with angle limits and motor speed

add enabled methods for motor and angular limits
use correct name to get joint
update copyright

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -198,14 +198,40 @@ void GodotPinJoint2D::solve(real_t p_step) {
 real_t GodotPinJoint2D::get_param(PhysicsServer2D::PinJointParam p_param) const {
-	if (p_param == PhysicsServer2D::PIN_JOINT_SOFTNESS) {
-		return softness;
+		switch(p_param) {
```
```suggestion
	switch(p_param) {
```

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -289,11 +292,56 @@ real_t PinJoint2D::get_softness() const {
 }
 
+void PinJoint2D::set_angular_limit_lower(real_t p_angular_limit_lower) {
+	angular_limit_lower = p_angular_limit_lower;
```
```suggestion
	if (angular_limit_lower == p_angular_limit_lower) {
		return;
	}
	angular_limit_lower = p_angular_limit_lower;
```
Avoid changes when value doesn't change

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -289,11 +292,56 @@ real_t PinJoint2D::get_softness() const {
+}
+
+void PinJoint2D::set_angular_limit_upper(real_t p_angular_limit_upper) {
+	angular_limit_upper = p_angular_limit_upper;
```
```suggestion
	if (angular_limit_upper == p_angular_limit_upper) {
		return;
	}
	angular_limit_upper = p_angular_limit_upper;
```

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -289,11 +292,56 @@ real_t PinJoint2D::get_softness() const {
+}
+
+void PinJoint2D::set_motor_target_velocity(real_t p_motor_target_velocity) {
+	motor_target_velocity = p_motor_target_velocity;
```
```suggestion
	if (motor_target_velocity == p_motor_target_velocity) {
		return;
	}
	motor_target_velocity = p_motor_target_velocity;
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -198,14 +198,66 @@ void GodotPinJoint2D::solve(real_t p_step) {
+	switch(p_param) {
+		case PhysicsServer2D::PIN_JOINT_SOFTNESS: {
+			softness = p_value;
+			break;
+		}
```
```suggestion
		} break;
```
Same everywhere

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -289,11 +295,98 @@ real_t PinJoint2D::get_softness() const {
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "angular_limit/lower", PROPERTY_HINT_RANGE, "-180,180,0.1,radians"), "set_angular_limit_lower", "get_angular_limit_lower");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "angular_limit/upper", PROPERTY_HINT_RANGE, "-180,180,0.1,radians"), "set_angular_limit_upper", "get_angular_limit_upper");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "motor/enable"), "set_enable_motor", "get_enable_motor");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "motor/target_velocity", PROPERTY_HINT_RANGE, "-200,200,0.01,or_greater,or_less,suffix:m/s"), "set_motor_target_velocity", "get_motor_target_velocity");
```
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "motor/target_velocity", PROPERTY_HINT_RANGE, U"-200,200,0.01,or_greater,or_less,radians,suffix:\u00B0/s"), "set_motor_target_velocity", "get_motor_target_velocity");
```
This is angular yes?

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -289,11 +295,98 @@ real_t PinJoint2D::get_softness() const {
+		PhysicsServer2D::get_singleton()->pin_joint_set_flag(get_joint(), PhysicsServer2D::PIN_JOINT_FLAG_ENABLE_MOTOR, enable_motor);
+	}
+}
+bool PinJoint2D::get_enable_motor() const {
```
```suggestion

bool PinJoint2D::get_enable_motor() const {
```
Leave space between functions

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -289,11 +295,98 @@ real_t PinJoint2D::get_softness() const {
+bool PinJoint2D::get_enable_motor() const {
+	return enable_motor;
+}
+void PinJoint2D::set_enable_limits(bool p_enable_limits) {
```
```suggestion

void PinJoint2D::set_enable_limits(bool p_enable_limits) {
```

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -289,11 +295,98 @@ real_t PinJoint2D::get_softness() const {
+		PhysicsServer2D::get_singleton()->pin_joint_set_flag(get_joint(), PhysicsServer2D::PIN_JOINT_FLAG_USE_LIMIT, enable_limits);
+	}
+}
+bool PinJoint2D::get_enable_limits() const {
```
```suggestion

bool PinJoint2D::get_enable_limits() const {
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -198,14 +198,66 @@ void GodotPinJoint2D::solve(real_t p_step) {
+	ERR_FAIL_V(0);
+}
+
+
```
```suggestion
```
No extra line

Ughuuu commented:
Took the motor part from here: https://github.com/slembcke/Chipmunk2D/blob/master/src/cpSimpleMotor.c
The joint limits part from here: https://github.com/slembcke/Chipmunk2D/blob/master/src/cpRotaryLimitJoint.c#L25

As I understood, the whole physics api now uses parts from chipmunk physics engine, which seems to be right.

Still need to write impl for calculating initial angle between objects.

AThousandShips commented:
We should add a reference to `COPYRIGHT.TXT` to cover that since it was forgotten when that file was added

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -155,6 +155,9 @@ bool GodotPinJoint2D::setup(real_t p_step) {
 
 	bias = delta * -(get_bias() == 0 ? space->get_constraint_bias() : get_bias()) * (1.0 / p_step);
 
+	// compute max impulse
```
```suggestion
	// Compute max impulse.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide).

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
 	if (B && dynamic_B) {
 		B->apply_impulse(P, rB);
 	}
+	// from https://github.com/slembcke/Chipmunk2D/blob/master/src/cpRotaryLimitJoint.c#L25
```
```suggestion
	// Taken from https://github.com/slembcke/Chipmunk2D/blob/master/src/cpRotaryLimitJoint.c#L25
```
Unsure if this is needed explicitly, but if so it should be a permalink or it won't be relevant in the future

Ughuuu replied:
Removing comment.

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+			pdist = angular_limit_lower - dist;
+		}
+		real_t error_bias = powf(1.0f - 0.1f, 60.0f);
+		// calculate bias velocity
```
```suggestion
		// Calculate bias velocity.
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
 		rel_vel = -vA;
 	}
 
+	// from https://github.com/slembcke/Chipmunk2D/blob/master/src/cpSimpleMotor.c#L45
+	// and https://github.com/slembcke/Chipmunk2D/blob/master/src/cpRotaryLimitJoint.c#L61
```
```suggestion
	// Taken from https://github.com/slembcke/Chipmunk2D/blob/master/src/cpSimpleMotor.c#L45
	// and https://github.com/slembcke/Chipmunk2D/blob/master/src/cpRotaryLimitJoint.c#L61
```
Same here for permalink

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+	// from https://github.com/slembcke/Chipmunk2D/blob/master/src/cpSimpleMotor.c#L45
+	// and https://github.com/slembcke/Chipmunk2D/blob/master/src/cpRotaryLimitJoint.c#L61
+	if ((enable_limits || enable_motor) && B) {
+		// compute relative rotational velocity
```
```suggestion
		// Compute relative rotational velocity.
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+		}
+		real_t j_max = jn_max;
+		
+		// compute normal impulse	
```
```suggestion
		// Compute normal impulse.
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+		// compute normal impulse	
+		real_t j = -(bias_velocity+wr)*i_sum;
+		real_t jOld = j_acc;
+		// only enable the limits if we have to
```
```suggestion
		// Only enable the limits if we have to.
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+	if (enable_limits && B) {
+		real_t dist = B->get_transform().get_rotation() - A->get_transform().get_rotation();
+		real_t pdist = 0.0f;
+		if(dist > angular_limit_upper) {
```
```suggestion
		if (dist > angular_limit_upper) {
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+		bias_velocity = CLAMP((-1.0f - powf(error_bias, p_step))*pdist/p_step, -get_max_bias(), get_max_bias());
+
+		// If the bias velocity is 0, the joint is not at a limit.
+		if(bias_velocity >= -CMP_EPSILON || bias_velocity <= CMP_EPSILON) {
```
```suggestion
		if (bias_velocity >= -CMP_EPSILON || bias_velocity <= CMP_EPSILON) {
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+		real_t jOld = j_acc;
+		// only enable the limits if we have to
+		if (enable_limits && is_joint_at_limit) {
+			if(bias_velocity < 0.0f){
```
```suggestion
			if (bias_velocity < 0.0f){
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+		}
+		real_t error_bias = powf(1.0f - 0.1f, 60.0f);
+		// calculate bias velocity
+		bias_velocity = CLAMP((-1.0f - powf(error_bias, p_step))*pdist/p_step, -get_max_bias(), get_max_bias());
```
```suggestion
		bias_velocity = CLAMP((-1.0 - powf(error_bias, p_step)) * pdist/p_step, -get_max_bias(), get_max_bias());
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+		} else if(dist < angular_limit_lower) {
+			pdist = angular_limit_lower - dist;
+		}
+		real_t error_bias = powf(1.0f - 0.1f, 60.0f);
```
```suggestion
		real_t error_bias = powf(1.0 - 0.1, 60.0);
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+	i_sum = 1.0f/(i_sum_local);
+	if (enable_limits && B) {
+		real_t dist = B->get_transform().get_rotation() - A->get_transform().get_rotation();
+		real_t pdist = 0.0f;
```
```suggestion
		real_t pdist = 0.0;
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+			if(bias_velocity < 0.0f){
+				j_acc = CLAMP(jOld + j, 0.0f, j_max);
+			} else {
+				j_acc = CLAMP(jOld + j, -j_max, 0.0f);
```
```suggestion
				j_acc = CLAMP(jOld + j, -j_max, 0.0);
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+		// only enable the limits if we have to
+		if (enable_limits && is_joint_at_limit) {
+			if(bias_velocity < 0.0f){
+				j_acc = CLAMP(jOld + j, 0.0f, j_max);
```
```suggestion
				j_acc = CLAMP(jOld + j, 0.0, j_max);
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+		real_t pdist = 0.0f;
+		if(dist > angular_limit_upper) {
+			pdist = angular_limit_upper - dist;
+		} else if(dist < angular_limit_lower) {
```
```suggestion
		} else if (dist < angular_limit_lower) {
```

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,21 +173,76 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
 	// compute relative velocity
 	Vector2 vA = A->get_linear_velocity() - custom_cross(rA - A->get_center_of_mass(), A->get_angular_velocity());
-
+	
```
```suggestion

```

Ughuuu commented:
Fixed the above issue, implemented feedback, review again and its ready for merging I think, from my side at least. Works also same with implementation in box2d, well, almost.

AThousandShips commented on COPYRIGHT.txt:
```diff
@@ -85,6 +85,11 @@ Copyright: 2001, Robert Penner
+Files: ./servers/physics_2d/godot_joints_2d.cpp
+Comment: Chipmunk2D Joint Constraints
+Copyright: Copyright (c) 2007 Scott Lembcke
+License:  MIT
```
```suggestion
License: MIT
```
My bad added extra space

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,12 +173,40 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+		} else if (dist < angular_limit_lower) {
+			pdist = dist - angular_limit_lower;
+		}
+		real_t error_bias = powf(1.0 - 0.15, 60.0);
```
```suggestion
		real_t error_bias = Math::pow(1.0 - 0.15, 60.0);
```
Use the built-in library for math

AThousandShips commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,12 +173,40 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
+		}
+		real_t error_bias = powf(1.0 - 0.15, 60.0);
+		// Calculate bias velocity.
+		bias_velocity = -CLAMP((-1.0 - powf(error_bias, p_step)) * pdist / p_step, -get_max_bias(), get_max_bias());
```
```suggestion
		bias_velocity = -CLAMP((-1.0 - Math::pow(error_bias, p_step)) * pdist / p_step, -get_max_bias(), get_max_bias());
```

rburing commented:
Node properties are more commonly called `*_enabled` than `enable_*`, which I think make sense because `enable_*` sounds more like a function. So for the `PinJoint2D` node I propose having a `motor_enabled` property with setter `set_motor_enabled` and getter `is_motor_enabled`, and an `angular_limit_enabled` property with setter `set_angular_limit_enabled` and getter `is_angular_limit_enabled`. I also propose renaming the `motor/target_velocity` property to `motor_target_velocity`, and similarly replacing `angular_limit/lower` by `angular_limit_lower` and `angular_limit/upper` by `angular_limit/upper`. This makes the properties easier to write in scripts, and we can still make  two groups in the inspector using the common prefixes, similar to how it is done in `RigidBody2D` here:

https://github.com/godotengine/godot/blob/5f1e56ff26be4070496aa51095b9ac2f2b4f4ed8/scene/2d/physics_body_2d.cpp#L1038-L1041

Ughuuu commented:
Thanks for the idea, very good naming scheme. I like it better. Updated. Also, angular_limit sounds a little bit weird. Should it be angular_limits?

AThousandShips commented on doc/classes/PinJoint2D.xml:
```diff
@@ -9,6 +9,21 @@
+			The maximum rotation. Only active if [member angular_limit_enabled] is [code]true[/code].
+		</member>
+		<member name="motor_enabled" type="bool" setter="set_motor_enabled" getter="is_motor_enabled" default="false">
+			When activated, a motor turns the hinge.
```
```suggestion
			When activated, a motor turns the pin.
```

AThousandShips commented on doc/classes/PhysicsServer2D.xml:
```diff
@@ -1160,6 +1177,21 @@
 			Constant to set/get a how much the bond of the pin joint can flex. The default value of this parameter is [code]0.0[/code].
 		</constant>
+		<constant name="PIN_JOINT_LIMIT_UPPER" value="1" enum="PinJointParam">
+			The maximum rotation across the Pin.
```
```suggestion
			The maximum rotation around the pin.
```

AThousandShips commented on doc/classes/PhysicsServer2D.xml:
```diff
@@ -1160,6 +1177,21 @@
+			The maximum rotation across the Pin.
+		</constant>
+		<constant name="PIN_JOINT_LIMIT_LOWER" value="2" enum="PinJointParam">
+			The minimum rotation across the Pin.
```
```suggestion
			The minimum rotation around the pin.
```

AThousandShips commented on doc/classes/PhysicsServer2D.xml:
```diff
@@ -1160,6 +1177,21 @@
+			Target speed for the motor. In radians per second.
+		</constant>
+		<constant name="PIN_JOINT_FLAG_ANGULAR_LIMIT_ENABLED" value="0" enum="PinJointFlag">
+			If [code]true[/code], the Pin has a maximum and a minimum rotation.
```
```suggestion
			If [code]true[/code], the pin has a maximum and a minimum rotation.
```

AThousandShips commented on doc/classes/PhysicsServer2D.xml:
```diff
@@ -1160,6 +1177,21 @@
+			If [code]true[/code], the Pin has a maximum and a minimum rotation.
+		</constant>
+		<constant name="PIN_JOINT_FLAG_MOTOR_ENABLED" value="1" enum="PinJointFlag">
+			If [code]true[/code], a motor turns the Pin.
```
```suggestion
			If [code]true[/code], a motor turns the pin.
```

AThousandShips commented on doc/classes/PinJoint2D.xml:
```diff
@@ -9,6 +9,21 @@
 	</tutorials>
 	<members>
+		<member name="angular_limit_enabled" type="bool" setter="set_angular_limit_enabled" getter="is_angular_limit_enabled" default="false">
+			If [code]true[/code], the pin maximum and minimum rotation, defined by [member angular_limit_lower] and [member angular_limit_upper] has effects.
```
```suggestion
			If [code]true[/code], the pin maximum and minimum rotation, defined by [member angular_limit_lower] and [member angular_limit_upper] are applied.
```

rburing commented:
> Also, angular_limit sounds a little bit weird. Should it be angular_limits?

Ah yes that does sound better, so a property `angular_limits_enabled` with getter `are_angular_limits_enabled` and setter `set_angular_limits_enabled`. Hopefully with that change the grouping still works.

Ughuuu commented:
Even though, looking again at the joints constraints for `physical_body_3d`, it's called `joint_constraints/angular_limit_enabled` there.

Ughuuu commented:
Here is project for other people to test:

https://github.com/godotengine/godot/assets/2369380/a4f0f310-5c61-4fdb-9bda-2fb9ecae01f7


[JointLimits.zip](https://github.com/godotengine/godot/files/12653224/JointLimits.zip)

rburing reviewed (approved):
Looks good to me, and will be helpful to integrate other 2D physics engines.

Also good to have Chipmunk2D in the credits where they should be.

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -289,11 +297,103 @@ real_t PinJoint2D::get_softness() const {
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "angular_limit_upper", PROPERTY_HINT_RANGE, "-180,180,0.1,radians"), "set_angular_limit_upper", "get_angular_limit_upper");
+	ADD_GROUP("Motor", "motor_");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "motor_enabled"), "set_motor_enabled", "is_motor_enabled");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "motor_target_velocity", PROPERTY_HINT_RANGE, U"-200,200,0.01,or_greater,or_less,suffix:radians,suffix:\u00B0/s"), "set_motor_target_velocity", "get_motor_target_velocity");
```
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "motor_target_velocity", PROPERTY_HINT_RANGE, U"-200,200,0.01,or_greater,or_less,radians,suffix:\u00B0/s"), "set_motor_target_velocity", "get_motor_target_velocity");
```

AThousandShips commented on scene/2d/joint_2d.h:
```diff
@@ -94,6 +99,17 @@ class PinJoint2D : public Joint2D {
+	void set_motor_target_velocity(real_t p_motor_target_velocity);
+	real_t get_motor_target_velocity() const;
+
+	void set_motor_enabled(bool motor_enabled);
```
```suggestion
	void set_motor_enabled(bool p_motor_enabled);
```

AThousandShips commented on scene/2d/joint_2d.h:
```diff
@@ -94,6 +99,17 @@ class PinJoint2D : public Joint2D {
+
+	void set_motor_enabled(bool motor_enabled);
+	bool is_motor_enabled() const;
+	void set_angular_limit_enabled(bool angular_limit_enabled);
```
```suggestion
	void set_angular_limit_enabled(bool p_angular_limit_enabled);
```

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -289,11 +297,103 @@ real_t PinJoint2D::get_softness() const {
+	angular_limit_lower = p_angular_limit_lower;
+	queue_redraw();
+	if (is_configured()) {
+		PhysicsServer2D::get_singleton()->pin_joint_set_param(get_joint(), PhysicsServer2D::PIN_JOINT_LIMIT_LOWER, p_angular_limit_lower);
```
```suggestion
		PhysicsServer2D::get_singleton()->pin_joint_set_param(get_rid(), PhysicsServer2D::PIN_JOINT_LIMIT_LOWER, p_angular_limit_lower);
```
After #80736

Ughuuu replied:
Uf, was wondering why build was failing now :D Thanks. This saves time.

akien-mga commented on COPYRIGHT.txt:
```diff
@@ -85,6 +85,11 @@ Copyright: 2001, Robert Penner
 
+Files: ./servers/physics_2d/godot_joints_2d.cpp
+Comment: Chipmunk2D Joint Constraints
+Copyright: Copyright (c) 2007 Scott Lembcke
+License: MIT
```
```suggestion
Copyright: 2007 Scott Lembcke
License: Expat
```

Ughuuu replied:
Will change tomorrow

akien-mga commented on servers/physics_2d/godot_joints_2d.cpp:
```diff
@@ -170,12 +173,40 @@ bool GodotPinJoint2D::pre_solve(real_t p_step) {
 	if (B && dynamic_B) {
 		B->apply_impulse(P, rB);
 	}
+	real_t i_sum_local = A->get_inv_inertia();
+	if (B) {
+		i_sum_local += B->get_inv_inertia();
+	}
+	i_sum = 1.0 / (i_sum_local);
+	if (angular_limit_enabled && B) {
+		Vector2 diff_vector = B->get_transform().get_origin() - A->get_transform().get_origin();
+		diff_vector = diff_vector.rotated(-initial_angle);
+		real_t dist = diff_vector.angle();
+		real_t pdist = 0.0;
+		if (dist > angular_limit_upper) {
+			pdist = dist - angular_limit_upper;
+		} else if (dist < angular_limit_lower) {
+			pdist = dist - angular_limit_lower;
+		}
+		real_t error_bias = Math::pow(1.0 - 0.15, 60.0);
+		// Calculate bias velocity.
+		bias_velocity = -CLAMP((-1.0 - Math::pow(error_bias, p_step)) * pdist / p_step, -get_max_bias(), get_max_bias());
+		// If the bias velocity is 0, the joint is not at a limit.
+		if (bias_velocity >= -CMP_EPSILON && bias_velocity <= CMP_EPSILON) {
+			j_acc = 0;
+			is_joint_at_limit = false;
+		} else {
+			is_joint_at_limit = true;
+		}
+	} else {
+		bias_velocity = 0.0;
+	}
```
If I understand correctly, it's this part, and the motor one below, which are derived from Chipmunk2D, right?

If so please add a comment in front of each of them to clarify where they derive from (with links), so it can be referred back easily in the future (and possibly updated if bugs are fixed upstream).

Ughuuu commented:
Thank you!

hackenshaw commented:
Why is it a soft limit? Is that intentional? When I think of limits I expect it to be a hard limit (like a hinge)

Ughuuu commented:
The code was taken from another physics engine that was initially used in Godot Physics 2d, chipmunk physics engine. And there it is implemented like that. I am not really an expert in 2d physics so working with an already implemented solution was easier.

hackenshaw commented:
This is definitely a good step. Thanks for your contribution. I'm also not a physics expert but I needed it for a client project. I'll take a look how it's implemented in box2d. Physics is always super finicky

Ughuuu commented:
In the box2d one the angle limits are not yet implemented as this was just merged.
In any case, here is the box2d repo: https://github.com/appsinacup/godot-box2d

weedleweedle commented:
> If you want hard limits, in box2d for eg. they have hard limits. I guess it depends on the physics engine and how it's implement? If not we can think of adding another parameter that specifies how hard/soft are the limits?

I think adding another parameter for the stiffness of the limits would help to provide both soft and hard limits

cody82 commented:
With which torque does it drive the joint?

Ughuuu commented:
For the angle limit or for the motor? Also the code is taken from the chipmunk engine, as can be seen, so I'm assuming it's similar to that.

cody82 commented:
I meant for the motor. Looks like it can apply unlimited torque.

Ughuuu commented:
I think so. If you want a different case I can make a proposal and add a maxTorque property maybe?

cody82 commented:
Right. But actually the motor does not work for me at all. It just does nothing. I have a wheel which is connected to another body but nothing happens if I set motor_enabled to true and motor_target_velocity to 100. What could be wrong?

Ughuuu commented:
> Here is project for other people to test:
> 
>  Screen.Recording.2023-09-18.at.22.23.26.mov 
> 
> [JointLimits.zip](https://github.com/godotengine/godot/files/12653224/JointLimits.zip)

Can you try this project that I tested this PR with and see if that works for you and try to work from that?

Ughuuu commented:
Also I'm not sure if this kind of help support discussion should happen here, as this PR is closed already. If you need anything, open a new godot proposal or ask on discord or on chat or on forum better. This way it won't notify the main devs also.
